dali/base/dasds.cpp

```diff
@@ -5697,116 +5735,225 @@ class CStoreHelper : implements IStoreHelper, public CInterface
         }
         return res;
     }
-    virtual void saveStore(IPropertyTree *root, unsigned *_newEdition)
+
+    enum class StoreFormat
+    {
+        XML,
+        BINARY
+    };
+    unsigned saveStoreToFile(const IPropertyTree *root, IFileIO *iFileIOTmpStore, StoreFormat format)
+    {
+        const bool isBinary = (format == StoreFormat::BINARY);
+        LOG(MCdebugProgress, "Saving store to %s", isBinary ? "binary" : "XML");
+        assertex(root);
+        assertex(iFileIOTmpStore);
+
+        // iFileIOTmpStore is passed in so that it can be used by the caller after the saveStoreToBinaryFile() call,
+        // for error processing and rename etc
+        IFile *iFile = iFileIOTmpStore->queryFile();
+        dbgassertex(iFile);
+        const char *fileName = iFile->queryFilename();
+        dbgassertex(fileName);
+        StringBuffer planeName;
+        constexpr size32_t bufferSize1mb = 0x100000;
+        size32_t bufferSize = bufferSize1mb;
+        if (findPlaneFromPath(fileName, planeName))
+            bufferSize = (size32_t)getPlaneAttributeValue(planeName, BlockedSequentialIO, bufferSize);
+
+        unsigned crc = 0;
+        if (isBinary)
+        {
+            Owned<ISerialOutputStream> serialStream = createSerialOutputStream(iFileIOTmpStore);
+            Owned<ICrcSerialOutputStream> crcSerialStream = createCrcOutputStream(serialStream);
+            Owned<IBufferedSerialOutputStream> bufOutStream = createBufferedOutputStream(crcSerialStream, bufferSize);
+            root->serializeToStream(*bufOutStream);
+            bufOutStream->flush();
+            bufOutStream.clear();
+            crcSerialStream->flush();
+            crc = crcSerialStream->queryCrc();
+            crcSerialStream.clear();
+            serialStream->flush();
+            serialStream.clear();
+        }
+        else
+        {
+            OwnedIFileIOStream fstream = createIOStream(iFileIOTmpStore);
+            Owned<ICrcIOStream> crcPipeStream = createCrcPipeStream(fstream);
+            Owned<IIOStream> ios = createBufferedIOStream(crcPipeStream, bufferSize);
+
+#ifdef _DEBUG
+            toXML(root, *ios); // formatted (default)
+#else
+            toXML(root, *ios, 0, 0);
```

Comment:
not for now: in a future PR it would be better if this code was commoned up, so both binary and the XML route used a IBufferedSerialOutputStream. Ideally toXML would be refactored to use IBufferedSerialOutputStream, but an alternative that would common-up code like this would be to introduce an adapter from IOStream->ISerialOutputStream, so that the majority of the code in this area would be common.